### ISSUE TITLE
When Shuffle is True, Get Data from the Workers of DataLoader's Iterator Randomly Rather Than One by One

### ISSUE THREAD

**Issue by lsy643** [pytorch/pytorch#78798]
### 🚀 The feature, motivation and pitch

When trying to get date from a `DataLoader`, we actually get data from the `data_queue` of the `_MultiProcessingDataLoaderIter` when `num_workers > 1`, which will create `num_workers` processes. And each process will output a batch every time. 

Because of the `index queue`,  `process 1` will always output a batch before `process 2`, even when Shuffle is set. Normally there is nothing wrong. However, in some situations in which data are stored in remote file systems and random augmentations for data are performed, `process 2` may get ready to output a batch before `process 1` but `process 2` will have to wait for `process 1`, therefore there might be some performance loss. 

  

cc @SsnL @VitalyFedyunin @ejguan @NivekT

**Comment by ejguan** [comment]
@lsy643
I don't think we would do this optimization because it will make the DataLoader completely non-deterministic in terms of the order to batches.

**Comment by lsy643** [comment]
@ejguan How about adding a parameter like `shuffle_workers: bool = False` to the `__init__` function of `DataLoader`.
The parameter `shuffle_workers` is valid only when `shuffle = True` and `num_workers > 1`, and only under such circumstances the data is generated from random workers otherwise the dataloader works as normal.  

If this is OK, I can make a PR to implement this.

**Comment by ejguan** [comment]
And, if you choose a larger `prefetch_factor`, you should be able to minimize such performance concern. Have you ever tested the performance in your case?

**Comment by ejguan** [comment]
If that's the case, I highly recommend you either use a custom solution or wait until `DataLoader2` in TorchData. I am reluctant to extend the feature inside `DataLoader` as the existing logic has already been really complicated. Without the order, it becomes hard for DataLoader to track the state of workers and when we should end iteration.
- For custom solution, you can create your own worker loop with index queue and result queue. You don't need the complicated logic inside DataLoader.
- For `DataLoader2` in the short future, in high-level, you should be able to create a own `MultiprocessingReadingService` to specify how to retrieve data from each worker process.

**Comment by lsy643** [comment]
@ejguan 
Thanks for your reply.  I will try DataLoader2 when it comes out.
